src/main.rs

```diff
@@ -12,29 +12,39 @@ mod slot_provider;
 use slot_provider::{mined_blocks::MinedBlocksSlotProvider, SlotProvider};
 
 async fn run_oracle_node(
-    price_provider: impl PriceProvider + 'static,
+    price_provider: impl PriceProvider + Clone + 'static,
     message_generator: MessageGenerator,
-    message_broadcaster: impl MessageBroadcaster + 'static,
+    message_broadcaster: impl MessageBroadcaster + Clone + 'static,
     slot_provider: impl SlotProvider,
 ) -> Result<()> {
     slot_provider
-        .run_for_every_slot(move |slot| -> Result<()> {
-            let price = price_provider
-                .get_price()
-                .wrap_err("Failed to get price data")?;
-            log::info!(
-                "Sucessfully obtained current Eth Price: {:?}",
-                price.value as f64 / PRECISION_FACTOR as f64
-            );
-            let oracle_message = message_generator
-                .generate_oracle_message(price, slot)
-                .wrap_err("Failed to generated signed price message")?;
-            log::info!("Sucessfully generated signed price message");
-            message_broadcaster
-                .broadcast(oracle_message)
-                .wrap_err("Failed to broadcast message")?;
-            Ok(())
-        })
+        .run_for_every_slot(
+            move |slot| -> Box<dyn futures::Future<Output = Result<()>> + Unpin> {
+                let message_broadcaster = message_broadcaster.clone();
+                let message_generator = message_generator.clone();
+                let price_provider = price_provider.clone();
+
+                Box::new(Box::pin(async move {
```

Comment:
Looking forward to playing with this a little to see if I can get the wrapping down but considering it's streams I'm not sure 😬 .

I like how it uses streams in general. Maybe async generators could help ease the pain.

Reply:
Yes for sure. Feel welcome to declutter this.
Beware though you might get sucked into an hours long rabbit hol. (which is what happened to me 😅).

I have worked mostly with streams in the past which is why I defaulted to using them here as well. But super interested to see how this can be done better / more succinctly.